pkg/controller/trainingjob_controller.go

```diff
@@ -0,0 +1,258 @@
+}
+
+// Run will set up the event handlers for trainingjob, as well as syncing
+// informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *TrainingJobController) Run(threadiness int, stopCh <-chan struct{}) error {
+	// TODO add a lock to ensure there is only one controller in the cluster
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	glog.Info("Starting trainingjob controller")
+	glog.Info("Starting to create custom resource definition")
+
+	if err := c.createCRD(); err != nil {
+		return fmt.Errorf("failed to create kind TrainingJob: %v", err)
+	}
+
+	glog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.trainingjobSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	glog.Info("Starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	glog.Info("Started workers")
+	<-stopCh
+	glog.Info("Shutting down workers")
+
+	return nil
+}
+
+func (c *TrainingJobController) createCRD() error {
+	crd := &extv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: paddlev1.CRDName(),
+		},
+		Spec: extv1beta1.CustomResourceDefinitionSpec{
+			Group:   paddlev1.CRDGroup,
+			Version: paddlev1.CRDVersion,
+			Scope:   extv1beta1.NamespaceScoped,
+			Names: extv1beta1.CustomResourceDefinitionNames{
+				Kind:       paddlev1.CRDKind,
+				Plural:     paddlev1.CRDKindPlural,
+				ShortNames: []string{paddlev1.CRDShortName},
+			},
+		},
+	}
+
+	_, err := c.ExtCli.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		glog.Errorf("Failed to create crd, error: %s", err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// enqueue takes a TrainingJob resource and converts it into a namespace/name
+// string which is then put onto the work queue.
+func (c *TrainingJobController) enqueue(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	glog.Infof("enqueue key: %v", key)
+	c.workqueue.AddRateLimited(key)
+}
+
+func (c *TrainingJobController) dequeue(obj interface{}) {
+	job, ok := obj.(*paddlev1.TrainingJob)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("type conversion error: %+v", obj))
+		return
+	}
+	key := job.Namespace + "/" + job.Name
+	glog.Infof("dequeue key: %v", key)
+	jobToDelete, ok := c.jobupdater.Load(key)
+	if !ok {
+		glog.Warningf("unsafe state. %s was never created but we received delete event", key)
+		return
+	}
+
+	oldtj := jobToDelete.(*updater.TrainingJobUpdater)
+	oldtj.Delete()
+	c.jobupdater.Delete(key)
+}
+
+func (c *TrainingJobController) runWorker() {
+	for c.processNestWorkItem() {
+	}
+}
+
+func (c *TrainingJobController) processNestWorkItem() bool {
```

Comment:
Looks like the name should be `processNextWorkItem`

Reply:
Yes, it's a typo here.